From f91f72bb3c73562aa890a5a37508dfb25f01f23f Mon Sep 17 00:00:00 2001
From: Nicholas Pridybailo <nikolai.pridybailo@edality.by>
Date: Fri, 31 May 2019 19:55:37 +0300
Subject: Add optee utils to trustbox apps list

optee_os, optee_client, optee_test, ntpm from Microsoft repositories
---
 configs/trustbox.cfg   | 15 ++++++++++-----
 include/repo.mk        |  9 +++++----
 packages/apps/Makefile | 41 ++++++++++++++++++++++++++++++++++++-----
 3 files changed, 51 insertions(+), 14 deletions(-)

diff --git a/configs/trustbox.cfg b/configs/trustbox.cfg
index 9cd5bf5..b30a10b 100644
--- a/configs/trustbox.cfg
+++ b/configs/trustbox.cfg
@@ -35,7 +35,7 @@ CONFIG_BUILD_PTPD=y
 CONFIG_BUILD_CRCONF=y
 CONFIG_BUILD_IPERF=y
 CONFIG_BUILD_OPENSTACK_NOVA=n
-CONFIG_BUILD_OPTEE=n
+CONFIG_BUILD_OPTEE=y
 CONFIG_BUILD_LIBPKCS11=n
 CONFIG_BUILD_SECURE_OBJ=n
 CONFIG_BUILD_EDGESCALE=n
@@ -101,7 +101,7 @@ fbdockerimgversion=18.04
 # component repository list for autobuild
 firmware_repo_list="atf u-boot qoriq-uefi-binary qoriq-fm-ucode mc-utils qoriq-mc-binary qoriq-qe-ucode qoriq-firmware-cortina qoriq-firmware-inphi qoriq-engine-pfe-bin ddr-phy-binary"
 linux_repo_list="linux cryptodev-linux lttng-modules rtl8723bu"
-apps_repo_list="restool flib fmlib fmc spc dpdk vpp pktgen-dpdk ovs-dpdk openssl gpp-aioptool cst aiopsl ptpd ceetm dce qbman_userspace eth-config crconf iperf optee_os optee_client optee_test secure_obj libpkcs11 qoriq-edgescale-eds qoriq-eds-kubelet qoriq-eds-bootstrap openstack-nova azure-iotedge"
+apps_repo_list="restool flib fmlib fmc spc dpdk vpp pktgen-dpdk ovs-dpdk openssl gpp-aioptool cst aiopsl ptpd ceetm dce qbman_userspace eth-config crconf iperf optee_os optee_client optee_test secure_obj libpkcs11 qoriq-edgescale-eds qoriq-eds-kubelet qoriq-eds-bootstrap openstack-nova azure-iotedge ftpm"
 
 # default git tree and branch
 default_uboot_tree=u-boot
@@ -181,9 +181,12 @@ iperf_repo_url=https://git.code.sf.net/p/iperf2/code.git
 iperf_repo_commit=1bc4619f874a14f51da2edfe800d23116b0bfa1e
 libpkcs11_repo_tag=LSDK-19.03
 secure_obj_repo_tag=LSDK-19.03
-optee_os_repo_tag=LSDK-19.03
-optee_client_repo_tag=LSDK-19.03
-optee_test_repo_tag=LSDK-19.03
+optee_os_repo_commit=fd17616cbc6508282fb37227dd21e5dfc9c49fe5
+optee_os_repo_url=https://github.com/ms-iot/optee_os.git
+optee_client_repo_commit=3745cedd13dd48672ea56126ee4506963b0d6fe4
+optee_client_repo_url=https://github.com/ms-iot/optee_client.git
+optee_test_repo_commit=c0e8678b55d399956da0bdce5b1f25c49839172c
+optee_test_repo_url=https://github.com/OP-TEE/optee_test.git
 openstack_nova_repo_url=https://github.com/openstack/nova
 openstack_nova_repo_branch=stable/pike
 qoriq_edgescale_eds_repo_url=https://github.com/nxp/qoriq-edgescale-eds.git
@@ -201,3 +204,5 @@ azure_iotedge_deb_name=iotedge_1.0.8~dev-1_arm64.deb
 azure_iotedge_deb_url=https://www.grapeboard.com/wp-content/pkgs/lsdk-1903/iotedge_1.0.8~dev-1_arm64.deb
 azure_libiothsm_deb_name=libiothsm-std_1.0.8~dev-1_arm64.deb
 azure_libiothsm_deb_url=https://www.grapeboard.com/wp-content/pkgs/lsdk-1903/libiothsm-std_1.0.8~dev-1_arm64.deb
+ftpm_repo_url=https://github.com/Microsoft/ms-tpm-20-ref.git
+ftpm_repo_commit=d979cae6db01ea2e57dbce92d700ce1d0edf78fa
diff --git a/include/repo.mk b/include/repo.mk
index b786404..2ec1513 100644
--- a/include/repo.mk
+++ b/include/repo.mk
@@ -14,13 +14,14 @@ define fetch-git-tree
 		commit=`grep -E "^$${tree}_repo_commit" $(FBDIR)/configs/$(CONFIGLIST) | cut -d= -f2` && \
 		repourl=`grep -E "^$${tree}_repo_url" $(FBDIR)/configs/$(CONFIGLIST) | cut -d= -f2` && \
 		if [ -z "$$repourl" ]; then repourl=$(GIT_REPOSITORY_URL)/$1; fi && \
-		if [ -n "$$tag" ]; then git clone --recurse-submodules $$repourl $1 && cd `realpath $(FBDIR)/packages/*/$1` && git checkout $$tag -b $$tag && cd -; \
-		elif [ -n "$$commit" ]; then git clone --recurse-submodules $$repourl $1 && cd `realpath $(FBDIR)/packages/*/$1` && git checkout $$commit -b $$commit && cd -; \
-		elif [ -n "$$branch" ]; then git clone --recurse-submodules $$repourl $1 -b $$branch; fi;\
+		if [ -n "$$tag" ]; then git clone --recurse-submodules $$repourl $1 && cd `realpath $(FBDIR)/packages/*/$1` && git checkout $$tag -b $$tag && \
+		git submodule init && git submodule update --init --recursive && cd -; \
+		elif [ -n "$$commit" ]; then git clone --recurse-submodules $$repourl $1 && cd `realpath $(FBDIR)/packages/*/$1` && git checkout $$commit -b $$commit && \
+		git submodule init && git submodule update --init --recursive && cd -; \
+		elif [ -n "$$branch" ]; then git clone --recurse-submodules $$repourl $1 -b $$branch && git submodule init && git submodule update --init --recursive; fi;\
 	fi
 endef
 
-
 define repo-update
 	@for tree in $2; do \
 	    echo -e "\nrepo: $$tree"; tree2=`echo $$tree | sed 's/-/_/g'`; \
diff --git a/packages/apps/Makefile b/packages/apps/Makefile
index 28537ed..421e7bf 100644
--- a/packages/apps/Makefile
+++ b/packages/apps/Makefile
@@ -294,15 +294,22 @@ ifeq ($(DESTARCH),arm64)
 		    flex-builder -c optee_os -m $$brd -a $(DESTARCH) -f $(CONFIGLIST); \
 		done;\
 	else\
-             if [ "ls1088ardb_pb" = $(MACHINE) ]; then brd=ls1088ardb; else brd=$(MACHINE); fi && \
-	     cd $(FBDIR)/packages/apps/optee_os && $(MAKE) CFG_ARM64_core=y PLATFORM=ls-$$brd ARCH=arm;\
+            if [ "ls1088ardb_pb" = $(MACHINE) ]; then \
+                brd=ls1088ardb; \
+            elif [ "trustbox" = $(MACHINE) ]; then \
+                 brd=ls1012grapeboard; \
+            else brd=$(MACHINE); fi && \
+	     cd $(FBDIR)/packages/apps/optee_os && $(MAKE) CFG_ARM64_core=y CFG_TEE_CORE_DEBUG=y CFG_TEE_CORE_LOG_LEVEL=2 PLATFORM=ls-$$brd ARCH=arm;\
 		$(CROSS_COMPILE)\objcopy -v -O binary out/arm-plat-ls/core/tee.elf out/arm-plat-ls/core/tee_$(MACHINE).bin;\
 	     if [ $(MACHINE) = ls1012afrwy ]; then \
 		mv out/arm-plat-ls/core/tee_$(MACHINE).bin out/arm-plat-ls/core/tee_$(MACHINE)_512mb.bin && \
 		$(MAKE) -j$(JOBS) CFG_ARM64_core=y PLATFORM=ls-ls1012afrwy ARCH=arm CFG_DRAM0_SIZE=0x40000000 && \
 		$(CROSS_COMPILE)\objcopy -v -O binary out/arm-plat-ls/core/tee.elf out/arm-plat-ls/core/tee_$(MACHINE).bin; \
              fi; \
-	fi && rm -f out/arm-plat-ls/core/tee.bin && $(call fbprint_d,"OPTEE_OS")
+	fi && \
+            if [ ! -d $(FBDIR)/build/firmware/optee ]; then \
+                mkdir -p $(FBDIR)/build/firmware/optee; fi && \
+	cp out/arm-plat-ls/core/tee_$(MACHINE).bin $(FBDIR)/build/firmware/optee/ && rm -f out/arm-plat-ls/core/tee.bin && $(call fbprint_d,"OPTEE_OS")
 endif
 else
 	@$(call fbprint_w,INFO: CONFIG_BUILD_OPTEE is not enabled by default in configs/$(CONFIGLIST))
@@ -328,10 +335,14 @@ optee_test:
 ifeq ($(CONFIG_BUILD_OPTEE), y)
 ifeq ($(DESTARCH),arm64)
 ifeq ($(CONFIG_BUILD_OPTEE), y)
-	@if [ ! -f $(DESTDIR)/lib/libteec.so.1.0 ]; then \
-		$(call fbprint_b,"Optee-client");\
+	@if [ ! -f $(DESTDIR)/lib/libteec.so.1.0 -o ! -d $(FBDIR)/packages/apps/optee_client ]; then \
+		$(call fbprint_n,"Optee-client required and not built yet. Building...");\
 		cd $(FBDIR) && flex-builder -c optee_client -a $(DESTARCH) -m $(MACHINE);\
 	fi;
+	@if [ ! -f $(FBDIR)/build/firmware/optee/tee_$(MACHINE).bin -o ! -d $(FBDIR)/packages/apps/optee_os ]; then \
+		$(call fbprint_n,"Optee-OS required and not built yet. Building...");\
+		cd $(FBDIR) && flex-builder -c optee_os -a $(DESTARCH) -m $(MACHINE);\
+	fi;
 	@$(call fbprint_b,"OPTEE_TEST")
 	@$(call fetch-git-tree,optee_test)
 	@cd $(FBDIR)/packages/apps/optee_client && $(MAKE) -j$(JOBS) ARCH=arm64;
@@ -348,6 +359,26 @@ endif
 endif
 
 
+.PHONY: ftpm
+ftpm:
+ifeq ($(CONFIG_BUILD_OPTEE), y)
+ifeq ($(DESTARCH),arm64)
+	@if [ ! -f $(FBDIR)/build/firmware/optee/tee_$(MACHINE).bin -o ! -d $(FBDIR)/packages/apps/optee_os ]; then \
+		$(call fbprint_n,"Optee-OS required and not built yet. Building...");\
+		cd $(FBDIR) && flex-builder -c optee_os -a $(DESTARCH) -m $(MACHINE);\
+	fi;
+	@$(call fbprint_b,"ftpm")
+	@$(call fetch-git-tree,ftpm)
+	@TA_DEV_KIT_DIR=$(FBDIR)/packages/apps/optee_os/out/arm-plat-ls/export-ta_arm64 \
+	TA_CPU=cortex-a53 CROSS_COMPILE=aarch64-linux-gnu- \
+	$(MAKE) -C $(FBDIR)/packages/apps/ftpm/Samples/ARM32-FirmwareTPM/optee_ta/fTPM \
+	O=$(FBDIR)/packages/apps/ftpm/build CFLAGS="$(CFLAGS) -Wno-expansion-to-defined"
+	@mkdir -p $(DESTDIR)/lib/optee_armtz
+	@cp $(FBDIR)/packages/apps/ftpm/build/bc50d971-d4c9-42c4-82cb-343fb7f37896.ta $(DESTDIR)/lib/optee_armtz
+endif
+endif
+
+
 .PHONY: libpkcs11
 libpkcs11:
 ifeq ($(CONFIG_BUILD_LIBPKCS11), y)
-- 
2.21.0

